### PR TITLE
Add a lint about using `clone` on `Copy` types

### LINKED ISSUE #594: lint on calling `clone` on `Copy` types.
The only reason `Copy` types should implement `Clone` is for generics, not for using the `clone` method on a concrete type.

### PR DESCRIPTION
**Pull request by mcarton**
Fix #594. Clippy had two instances of this.
Also run `rustfmt` because `str/method.rs` was really badly formatted since #540. Tell me if you’d prefer separate PRs.

### CODE REVIEW

File: src/methods.rs [hunk 1]
```diff
@@ -440,6 +455,18 @@ fn lint_or_fun_call(cx: &LateContext, expr: &Expr, name: &str, args: &[P<Expr>])
     }
 }
 
+/// Checks for the `CLONE_ON_COPY` lint.
+fn lint_clone_on_copy(cx: &LateContext, expr: &Expr, name: &str, args: &[P<Expr>]) {
+    if args.len() == 1 && name == "clone" {
+        let ty = cx.tcx.expr_ty(expr);
+        let parameter_environment = cx.tcx.empty_parameter_environment();
+
+        if !ty.moves_by_default(&parameter_environment, expr.span) {
```
Comment (oli-obk) [inline src/methods.rs]: Does this work for generic functions where the type is a generic parameter that impls copy?

Comment (Manishearth) [inline src/methods.rs]: Why not just look for `#[automatically_derived]` impls for the same base type?

Comment (mcarton) [inline src/methods.rs]: @oli-obk now it does :smile:
@Manishearth I’m not sure what you mean. This has nothing to do with deriving traits.

Comment (Manishearth) [inline src/methods.rs]: Oh, duh, Copy is special

**Comment by mcarton** [comment]
I just rebased. Is there a reason not to merge this one yet? I don’t mind but since there is a lot of modifications here and there, it diverged quite a lot.
@oli-obk’s comment has been taken care of.

**Comment by Manishearth** [comment]
Sorry about the delay. Stuff falls through the cracks sometimes :smile:

**Comment by mcarton** [comment]
No problem :smile:

**Pull request merged**